### PR TITLE
Fix/subscribe offset

### PR DESCRIPTION
**Pull request by wonesy**
This PR addresses several issues:

1. There was one undefined response code 19 
image
2. When subscribing to FIRST or LAST on a fresh stream, it would error out because of NoOffset. This handles the error and resets offset 0. (Question: should we store the offset at this point?)

### CODE REVIEW

**Comment by wonesy** [comment]
By the way this fixes the test that is blocking the PR #22 @Gsantomaggio

File: rstream/exceptions.py
Comment (qweeze) [inline rstream/exceptions.py:85]: nit: maybe `OffsetNotFound` would be a slightly better name? As in dotnet [client](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/main/RabbitMQ.Stream.Client/ResponseCode.cs#L27)

**Review by qweeze — APPROVED** [review]

**Comment by qweeze** [comment]
> should we store the offset at this point?

I'm not sure, but I guess we should do what other clients do :slightly_smiling_face: 
Seems that go client doesn't store offset when handling this [situation](https://github.com/rabbitmq/rabbitmq-stream-go-client/blob/main/pkg/stream/client.go#L719). And dotnet client doesn't even handle this exception explicitly. Probably @Gsantomaggio knows better

**Comment by wonesy** [comment]
Ok I think we can safely follow the official clients' implementations at this point and not save it

**Comment by Gsantomaggio** [comment]
@wonesy @qweeze 

we should change a couple of things.

Change consumer behaviour 
---
remove offset query from [the consumer](https://github.com/qweeze/rstream/pull/25/files#diff-40e4a1eedc170b2d93b57857ad9e0b9e2440804e2019cf03ea2c16f54b391ee4R127-R131)

just remote everything:
```diff
- if offset_type in (OffsetType.LAST, OffsetType.NEXT):
- offset = await self.query_offset(stream, reference)
-            try:
-                offset = await self.query_offset(stream, reference)
-            except exceptions.OffsetNotFound:
-                offset = 0
```

if the user wants to consume from `FIRST` or `LAST ` we don't have to query the offset. 

Expose the query offset to the consumer level.
---

The interaction would be like:
```python
 consumer = Consumer(
       ...
    )

....

stream = "mystream"
subscriber_name = "my_sub"

last_stored = await consumer.query_offset(stream, subscriber_name)

await consumer.start()
await consumer.subscribe(stream, on_message,
                             decoder=amqp_decoder,
                             subscriber_name= subscriber_name,
                             offset_type=OffsetType.OFFSET,  
                             offset=last_stored)
```


Note: Only if the `subscriber_name` is != "" you can execute the query


this is how .NET client works:

```csharp
  var offset = await system.QueryOffset(Reference, stream);
   var consumerWithOffset = await system.CreateConsumer(
    new ConsumerConfig
      {
         Reference = Reference,
         Stream = stream,
         OffsetSpec = new OffsetTypeOffset(offset), 
```

it is up to the user to decide what to do with `NoOffsetFound` error.

At some point, we could add some smart layer as the go client does with `GetLastConsumed` or stuff like that

**Comment by wonesy** [comment]
I see, so the higher-level offset types should work without any specified offset value.

```python
async def test_offset_type_last(stream: str, consumer: Consumer, producer: Producer) -> None:
    messages = [str(i).encode() for i in range(1, 11)]
    await producer.publish_batch(stream, messages)

    captured: list[bytes] = []
    await consumer.subscribe(
        stream,
        callback=captured.append,
        offset_type=OffsetType.LAST,
        subscriber_name="test-subscriber",
    )

    await producer.publish(stream, b"11")

    await wait_for(lambda: len(captured) >= 2)
    assert captured == [b"10", b"11"]
```

I'm trying to refactor a test which breaks when I remove that block as you suggested above. I just want check that this new test *should* work (it doesn't :))

**Comment by qweeze** [comment]
@wonesy Apparently this test doesn't work because rabbit sends not only the last one, but more messages from the first batch (probably due to chunking?). I tried it with different sizes of `messages`, for example if I do
```python
messages = [str(i).encode() for i in range(1, 10_000)]
```
then `len(captured)` is 1841.

Even if that's how it is supposed to work, it seems such behavior can be unexpected for users. So maybe we should drop extra messages on client side. @Gsantomaggio how do other clients deal with this?

**Comment by wonesy** [comment]
From what I can tell, the go client deals with all incoming messages [https://github.com/rabbitmq/rabbitmq-stream-go-client/blob/main/pkg/stream/client.go#L767-L793](here)

Doesn't look like this is considered a problem, but I agree it's unexpected from my end.

**Comment by Gsantomaggio** [comment]
Ok let's try to recap:
```python
class OffsetType(int, enum.Enum):
    FIRST = 1
    LAST = 2
    NEXT = 3
    OFFSET = 4
    TIMESTAMP = 5
```

`FIRST`,`NEXT`,`LAST` are referred to `chuck`

In one `chunk` there could be more messages

So when you ask for the `LAST` it means `last chuck`.

if you ask for the `OFFSET` it means `the chuck that contains the offset` in this case the client has to manually filter the previous messages:
 - [where .NET client filters](https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/blob/main/RabbitMQ.Stream.Client/Consumer.cs#L88-L93) 
- [Where the Go filters](https://github.com/rabbitmq/rabbitmq-stream-go-client/blob/8fbdf9dda96c13bb6c14aeeff1064260cdab4bd3/pkg/stream/server_frame.go#L396-L398)

with `FIRST`,`NEXT`,`LAST` no client-side filters are needed.

Hope it is clear.

( btw you can find me on [slack](https://rabbitmq-slack.herokuapp.com/) user: `gsantomaggio` channel if you want to have a chat about that).

Thank you

**Comment by Gsantomaggio** [comment]
> it's a common use case to get only the last message, 

you are right, but the basic protocol doesn't provide it. This is why the go client has [LastConsumsed() ](https://github.com/rabbitmq/rabbitmq-stream-go-client/blob/main/pkg/stream/client.go#L715).

We could do the same in this client, but in this specific PR I would stay close to the protocol. 
Maybe later we can add some smart flag, something like:

```python
class OffsetType(int, enum.Enum):
    FIRST = 1
    LAST = 2
    NEXT = 3
    OFFSET = 4
    TIMESTAMP = 5
    LAST_MESSAGE_STORED= 20 ### in this case we can  automatically retrieve the last stored offset and start consuming from it
```

**Comment by wonesy** [comment]
Thanks man, super helpful. I pushed another commit after this conversation.

**Comment by qweeze** [comment]
Now that we've sorted this out with offsets I think we should also fix filtering [here](https://github.com/qweeze/rstream/blob/master/rstream/consumer.py#L202) and make it the same as in Go and .NET clients

File: tests/test_consumer.py
Comment (qweeze) [inline tests/test_consumer.py:89]: we can do this to avoid `sleep(0.1)`
```suggestion
    await wait_for(lambda: captured[-1] == b"4999")
```

File: tests/test_consumer.py
Comment (qweeze) [inline tests/test_consumer.py:140]: Seems this test will pass for both LAST and FIRST. I'm curious if we could write a test that would reliably pass only for LAST

Comment (wonesy) [reply tests/test_consumer.py:139]: Yeah you're right this test doesn't belong here any longer. The one assert that seems to actually test LAST might be here https://github.com/qweeze/rstream/pull/25/files#diff-b6b23775d81e14e919519e377f7d4c6f9bcd28db05bfbd5645492a9a2f18cd81R90

Comment (wonesy) [reply tests/test_consumer.py:139]: Is there something with query_offset we can try

**Comment by Gsantomaggio** [comment]
@wonesy just tried your [commit](https://github.com/qweeze/rstream/pull/25/commits/bf82eb5521c6718edb75e8a01a9ab5a8c5aa074f):

```

tests/test_consumer.py:123: IndexError
======================================================================= short test summary info =======================================================================
FAILED tests/test_consumer.py::test_offset_type_last - IndexError: list index out of range
FAILED tests/test_consumer.py::test_offset_type_timestamp - IndexError: list index out of range
````

**Comment by wonesy** [comment]
try now

**Comment by Gsantomaggio** [comment]
```

/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py:492: TimeoutError
======================================================================= short test summary info =======================================================================
FAILED tests/test_consumer.py::test_offset_type_timestamp - asyncio.exceptions.TimeoutError
==================================================================== 1 failed, 52 passed in 3.02s =====================================================================
```

**Comment by qweeze** [comment]
I have TimeoutError 3 of 5 times now, probably there's a race condition in `wait_for`

**Comment by wonesy** [comment]
Intersting.

What if this diff is applied?

```
diff --git a/tests/test_consumer.py b/tests/test_consumer.py
index 4b42bc7..9cf41a6 100644
--- a/tests/test_consumer.py
+++ b/tests/test_consumer.py
@@ -101,7 +101,7 @@ async def test_offset_type_timestamp(stream: str, consumer: Consumer, producer:
     # create subscriber and declare to server, unsub immediately
     await consumer.subscribe(
         stream,
-        callback=captured.append,
+        callback=lambda: None,
         subscriber_name=subscriber_name
     )
:
```

**Comment by wonesy** [comment]
I always see it pass on my machine so i'm not sure which aspect is failing

**Comment by qweeze** [comment]
On my machine the issue is that some messages of the first chunk are captured between `subscribe` and `unsubscribe`. And the tests always passes when I add `captured.clear()` on line 108.

**Comment by wonesy** [comment]
That makes sense, I also experimented with making the first callback `lambda: None`

**Comment by qweeze** [comment]
Btw, what's the purpose of this subsribe-unsubscribe [thing](https://github.com/qweeze/rstream/pull/25/files#diff-b6b23775d81e14e919519e377f7d4c6f9bcd28db05bfbd5645492a9a2f18cd81R101-R107)?

File: rstream/consumer.py
Comment (qweeze) [inline rstream/consumer.py:203]: I understand that this can be annoying, but python's convention is to use snake_case for var names :)

Comment (wonesy) [reply rstream/consumer.py]: I write go a lot, my brain is not as fast at the language switch as i'd like it to be

**Comment by wonesy** [comment]
This is the one gents. I think I understand what's going on

> Btw, what's the purpose of this subsribe-unsubscribe [thing](https://github.com/qweeze/rstream/pull/25/files#diff-b6b23775d81e14e919519e377f7d4c6f9bcd28db05bfbd5645492a9a2f18cd81R101-R107)?

It was my misunderstanding of the internals

**Comment by Gsantomaggio** [comment]
@wonesy @qweeze what if we drop the time stamp test for the moment then open another PR.

I realized  (thanks @wonesy :) )  that I have to add this test in Go,.NEt and Rust.

We can "copy" from the[ java client](https://github.com/rabbitmq/rabbitmq-stream-java-client/blob/main/src/test/java/com/rabbitmq/stream/impl/OffsetTest.java#L352-L391) that is the only one that implements the test ATM

So we can unblock the CI and Merge this PR.

**Comment by wonesy** [comment]
no objection on my end

**Comment by qweeze** [comment]
Fine by me. @wonesy can I merge this PR now?

**Comment by wonesy** [comment]
Go for it

**Pull request merged**
